src/bitcoind.cpp

```diff
@@ -105,24 +187,34 @@ static bool AppInit(int argc, char* argv[])
             // InitError will have been called with detailed error, which ends up on console
             return false;
         }
-        if (args.GetBoolArg("-daemon", false)) {
-#if HAVE_DECL_DAEMON
-#if defined(MAC_OSX)
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
-#endif
+        if (args.GetBoolArg("-daemon", DEFAULT_DAEMON) || args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
+#if HAVE_DECL_FORK
             tfm::format(std::cout, PACKAGE_NAME " starting\n");
 
             // Daemonize
-            if (daemon(1, 0)) { // don't chdir (1), do close FDs (0)
-                return InitError(Untranslated(strprintf("daemon() failed: %s\n", strerror(errno))));
+            switch (fork_daemon(1, 0, daemon_ep)) { // don't chdir (1), do close FDs (0)
+                case 0: // Child: continue.
+                    // If -daemonwait is not enabled, immediately send a success token the parent.
+                    if (!args.GetBoolArg("-daemonwait", DEFAULT_DAEMONWAIT)) {
+                        daemon_ep.TokenWrite(1);
+                        daemon_ep.Close();
+                    }
+                    break;
+                case -1: // Error happened.
+                    return InitError(Untranslated(strprintf("fork_daemon() failed: %s\n", strerror(errno))));
+                default: { // Parent: wait and exit.
+                        int token = daemon_ep.TokenRead();
+                        if (token == true) { // fRet == true, Success
```

Comment:
nit:
```suggestion
                        if (token == 1) { // fRet == 1, Success
```
(probably the token variable could be eliminated by calling `TokenRead` directly in the if condition.)

Reply:
I prefer having a variable here. Having a read as a side effect of an if() clause is kind of meh.
Will change the `true`.